### PR TITLE
Rwa 441

### PR DESCRIPTION
**Pull request by lewis150193**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-244

### Change description ###
Idam pact test

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x ] No
```

### CODE REVIEW

**Review by iguannago — APPROVED** [review]

File: build.gradle [hunk 1]
```diff
@@ -19,6 +19,7 @@ plugins {
   id 'org.sonarqube' version '3.0'
   id 'info.solidsoft.pitest' version '1.5.2'
   id 'io.freefair.lombok' version '5.1.0'
+  id 'au.com.dius.pact' version '4.0.4'
```
Comment (BenNzewi) [inline build.gradle:22]: Please reference the following ia-case-api branch "ConsumerContractTests". It's the most up to date PACT change for case-api.

version should be -> id 'au.com.dius.pact' version '4.1.0'

File: src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/CcdCasePactTest.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
+package uk.gov.hmcts.reform.wataskconfigurationapi;
```
Comment (BenNzewi) [inline src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/CcdCasePactTest.java:1]: Should this be under in a CCD PACT Test class?

File: src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/IdamConsumerTestForPostToken.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.TreeMap;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class IdamConsumerTestForPostToken extends SpringBootContractBaseTest {
+
+    private static final String IDAM_OPENID_TOKEN_URL = "/o/token";
+
+    @Pact(provider = "Idam_api", consumer = "wa_task_management_api")
```
Comment (BenNzewi) [inline src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/IdamConsumerTestForPostToken.java:28]: It appears we are referring to an out of date implementation of case-api. I spoke to Satya and we should use the implementation on the following ia-case-api branch "ConsumerContractTests".

Please reference the following ia-case-api branch "ConsumerContractTests".
For example;
In @Pact(provider = "Idam_api", consumer = "wa_task_management_api"), Idam_api should be idamApi_oidc

File: src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/IdamConsumerTestForPostUserInfo.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Map;
+import java.util.TreeMap;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class IdamConsumerTestForPostUserInfo extends SpringBootContractBaseTest {
+
+    private static final String IDAM_OPENID_USERINFO_URL = "/o/userinfo";
+
+    @Pact(provider = "Idam_api", consumer = "wa_task_management_api")
```
Comment (BenNzewi) [inline src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/IdamConsumerTestForPostUserInfo.java:30]: Please reference the following ia-case-api branch "ConsumerContractTests".
For example;
In @Pact(provider = "Idam_api", consumer = "wa_task_management_api"), Idam_api should be idamApi_oidc

File: src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/ccdpacttest/CcdCasePactTest.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    public void should_post_to_token_endpoint_and_receive_access_token_with_200_response(MockServer mockServer)
+        throws JSONException {
+        String actualResponseBody =
+            SerenityRest
+                .given()
+                .contentType(ContentType.URLENC)
+                .log().all(true)
+                .get(mockServer.getUrl() + CCD_CASE_URL)
+                .then()
+                .extract().asString();
```
Comment (indvamshi) [inline src/contractTest/java/uk/gov/hmcts/reform/wataskconfigurationapi/ccdpacttest/CcdCasePactTest.java:57]: can we verify for 200 status code. Same in other scenarios as well.

File: build.gradle [hunk 6]
```diff
@@ -311,6 +342,30 @@ dependencies {
...
+  logger.lifecycle("Runs pact Tests")
+  testClassesDirs = sourceSets.contractTest.output.classesDirs
+  classpath = sourceSets.contractTest.runtimeClasspath
+
+}
+
+runAndPublishConsumerPactTests.finalizedBy pactPublish
+
+pact {
+  publish {
```
Comment (alexbaeza) [inline build.gradle:359]: There's some config missing here used by the new version see this one for an example:
https://github.com/hmcts/wa-task-management-api/pull/169

**Review by alexbaeza — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 7]
```diff
@@ -119,6 +129,18 @@ task tests {
...
+task contract(type: Test) {
+  description = "Runs the consumer Pact tests"
+  group = 'Verification'
+  useJUnitPlatform()
+  testClassesDirs = sourceSets.contractTest.output.classesDirs
+  classpath = sourceSets.contractTest.runtimeClasspath
+}
+
+project.ext {
+  pacticipant = 'wa_taskConfigurationApi'
```
Comment (alexbaeza) [inline build.gradle:141]:
```suggestion
  pacticipant = 'wa_task_configuration_api'
```

**Review by alexbaeza — APPROVED** [review]

**Pull request merged**